### DISCUSSION TITLE
0085 XLS-85d: Token Escrow — Token-Enabled Escrows

### DISCUSSION THREAD

**Discussion by dangell7** [XRPLF/XRPL-Standards#248] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by xVet** [comment]
@dangell7 well done Denis, this is a highly requested feature - as well as Ed and Richard ❤️

Quick question, the transfer fee you mention and carry over to the escrow is the issuer transfer fee we know from IOUs and not a new fee, correct?

**Reply by dangell7** [reply]
Correct. The `TransferRate` is not something the user supplies, its taken directly from the Issuer `AccountRoot`.

**Reply by xVet** [reply]
> @dangell7 well done Denis, this is a highly requested feature - as well as Ed and Richard ❤️
> 
> Quick question, the transfer fee you mention and carry over to the escrow is the issuer transfer fee we know from IOUs and not a new fee, correct?

In addition, is the flag able to be set even if tokens have been issued already to allow backwards compatibility or is this only for new issued tokens?

**Reply by dangell7** [reply]
I don't think this works if we don;t allow already issued tokens to use it. I say allow to be set and unset at any time. Unsetting would remove the ability to create new escrows but should not effect already escrowed tokens.

**Reply by xVet** [reply]
One last thing - should the transfer rate not also be described as "(Optional)" ? Not all tokens have it and therefore not all escrows will.

**Reply by dangell7** [reply]
It's an optional field indeed. On Token escrows if there is no transfer rate we were/are storing the parity rate. So right now every Token Escrow will have a `TransferRate` but I can just as well remove that as its not 100% necessary. Thanks for pointing that out.

**Reply by xVet** [reply]
Thanks! Appreciate the time you put into this 🙏

**Comment by KhaledElawadi** [comment]
Like it !

Would prefer to add more programmability/functionality to the Escrow protocol though by enabling direct services such as  "Escrow to AMM" and "Escrow to Listed Accounts".

The former would be useful for things like programmed token release/distribution to AMM and timed airdrop/incentives/rewards to AMM LP'ers and the latter for scheduled push payments to individuals/multiple accounts to accommodate things like Direct Debit and Loan/Mortgage repayments etc.

**Comment by tequdev** [comment]
I don't know much about MPToken, but could this support MPToken as well?

**Reply by mvadari** [reply]
That would probably have to be a separate spec.

**Reply by KhaledElawadi** [reply]
How about LP tokens?

Could they be escrowed too?

**Comment by dangell7** [comment]
This spec has been updated to include MPT tokens. It also outlines the nuance between the two types of tokens, and how each Escrow transaction will handle IOU and MPT differently.
